core/src/ast/validity.rs

```diff
@@ -1,43 +1,13 @@
-use super::{Ident, Path, TypeName};
+use super::TypeName;
 
+/// Errors that must happen during AST validation
 #[cfg_attr(feature = "displaydoc", derive(displaydoc::Display))]
 #[derive(Debug, Clone)]
 #[non_exhaustive]
 pub enum ValidityError {
-    /// An oqaue type crosses the FFI boundary as a value.
-    #[cfg_attr(
-        feature = "displaydoc",
-        displaydoc("An opaque type crossed the FFI boundary as a value: {0}")
-    )]
-    OpaqueAsValue(TypeName),
-    /// A non-oquare zero-sized struct or enum has been defined.
-    #[cfg_attr(
-        feature = "displaydoc",
-        displaydoc("A non-opaque zero-sized struct or enum has been defined: {0}")
-    )]
-    NonOpaqueZST(Path),
-    /// A non-opaque type was found behind a `Box` or reference.
-    #[cfg_attr(
-        feature = "displaydoc",
-        displaydoc(
-            "A non-opaque type was found behind a Box or reference, these can \
-               only be handled by-move as they get converted at the FFI boundary: {0}"
-        )
-    )]
-    NonOpaqueBehindRef(TypeName),
-    /// An Option was found in an unsupported location
-    #[cfg_attr(
-        feature = "displaydoc",
-        displaydoc("An Option was found in an unsupported location: {0}")
-    )]
-    OptionNotInReturnPosition(TypeName),
-    /// A Result was found in an unsupported location
-    #[cfg_attr(
-        feature = "displaydoc",
-        displaydoc("A Result was found in an unsupported location: {0}")
-    )]
-    ResultNotInReturnPosition(TypeName),
     /// A return type contains elided lifetimes.
+    ///
+    /// This is necessary during AST validation since HIR lifetime lowering will fail otherwise.
```

Comment:
shouldn't we have HIR lowering fail? It's weird to have a single-variant enum and handle all other errors differently

Reply:
we should, it's more complicated to move this check over so I'm not doing so just yet